app/src/ui/lib/ui-activity-monitor.ts

```diff
@@ -0,0 +1,138 @@
+import { Emitter, Disposable } from 'event-kit'
+import { ipcRenderer } from 'electron'
+
+/**
+ * Describes the interface of a UI activity monitor.
+ *
+ * We define a UI activity monitor as an object which can
+ * be used to be notified whenever the user interacts with the
+ * application.
+ */
+export interface IUiActivityMonitor {
+  /**
+   * Subscribe to get notified when the user interacts
+   * with the application. The first, and only argument
+   * to the event handler will be a value indicating the
+   * kind of action detected (mouse/pointer, keyboard etc).
+   *
+   * @return A disposable object which, when disposed will
+   *         terminate the subscription and prevent any
+   *         further calls to the handler.
+   */
+  onActivity(handler: (kind: UiActivityKind) => void): Disposable
+}
+
+/**
+ * A value indicating a user activity type (mouse/pointer, keyboard etc).
+ */
+export type UiActivityKind = 'pointer' | 'keyboard' | 'menu'
+
+/**
+ * UI Activity monitor for user interactions within GitHub Desktop.
+ */
+export class UiActivityMonitor implements IUiActivityMonitor {
+  private readonly emitter = new Emitter()
+  private subscriberCount = 0
+
+  /**
+   * Subscribe to get notified when the user interacts
+   * with the application. The first, and only argument
+   * to the event handler will be a value indicating the
+   * kind of action detected (mouse/pointer, keyboard etc).
+   *
+   * @return A disposable object which, when disposed will
+   *         terminate the subscription and prevent any
+   *         further calls to the handler.
+   */
+  public onActivity(handler: (kind: UiActivityKind) => void): Disposable {
+    const emitterDisposable = this.emitter.on('activity', handler)
+
+    if (this.subscriberCount++ === 0) {
+      this.startTracking()
+    }
+
+    return new Disposable(() => {
+      if (--this.subscriberCount === 0) {
+        this.stopTracking()
+      }
+
+      emitterDisposable.dispose()
+    })
+  }
+
+  private emit(kind: UiActivityKind) {
+    this.emitter.emit('activity', kind)
+  }
+
+  private startTracking() {
+    document.addEventListener('mousedown', this.onMouseDown)
+    document.addEventListener('keydown', this.onKeyDown)
+
+    ipcRenderer.on('menu-event', this.onMenuEvent)
+  }
+
+  private stopTracking() {
+    document.removeEventListener('mousedown', this.onMouseDown)
+    document.removeEventListener('keydown', this.onKeyDown)
+
+    ipcRenderer.removeListener('menu-event', this.onMenuEvent)
+  }
+
+  private onMouseDown = (e: MouseEvent) => {
+    if (e.target === null || !(e.target instanceof HTMLElement)) {
+      return
+    }
+
+    if (!isInteractionTarget(e.target)) {
+      return
+    }
+
+    this.emit('pointer')
+  }
+
+  private onKeyDown = (e: KeyboardEvent) => {
+    this.emit('keyboard')
```

Comment:
For my education only:

Does this mean we count any keyboard press as in-app activity? For instance, if I open the app and press <kbd>J</kbd>, will that count as activity in the app?

I see the mouse event is being inspected to ensure it's not just a random click somewhere in the app. I assume it's not possible to do the same thing with a [keyboard event](https://reactjs.org/docs/events.html#keyboard-events) since it doesn't include any properties that would help there.

Reply:
> Does this mean we count any keyboard press as in-app activity? For instance, if I open the app and press J, will that count as activity in the app?

Yes, see the condition in the PR body. Any keypress is treated as an interaction. I'm not too concerned about false-positives here and it's tricky to figure out if a keyboard event did anything since some components will subscribe to keyboard events on random events (such as the List component).